packages/flutter/lib/src/material/tabs.dart

```diff
@@ -760,543 +352,478 @@ class TabBar<T> extends Scrollable implements AppBarBottomWidget {
+    final ScrollableState viewport = viewportKey.currentState;
+    final double left = _currentIndex > 0 ? _tabCenteredScrollOffset(viewport, _currentIndex - 1) : null;
+    final double middle = _tabCenteredScrollOffset(viewport, _currentIndex);
+    final double right = _currentIndex < maxTabIndex ? _tabCenteredScrollOffset(viewport, _currentIndex + 1) : null;
+
+    final double index = _controller.index.toDouble();
+    final double value = _controller.animation.value;
+    double offset;
+    if (value == index - 1.0)
+      offset = left ?? middle;
+    else if (value == index + 1.0)
+      offset = right ?? middle;
+    else if (value == index)
+       offset = middle;
+    else if (value < index)
+      offset = left == null ? middle : lerpDouble(middle, left, index - value);
+    else
+      offset = right == null ? middle : lerpDouble(middle, right, value - index);
 
-  @override
-  ExtentScrollBehavior createScrollBehavior() {
-    return new _TabsScrollBehavior()
-      ..isScrollable = config.isScrollable;
+    viewport.scrollTo(offset);
   }
 
-  @override
-  _TabsScrollBehavior get scrollBehavior => super.scrollBehavior;
-
-  double _centeredTabScrollOffset(int tabIndex) {
-    double viewportWidth = scrollBehavior.containerExtent;
-    Rect tabRect = _tabRect(tabIndex);
-    return (tabRect.left + tabRect.width / 2.0 - viewportWidth / 2.0)
-      .clamp(scrollBehavior.minScrollOffset, scrollBehavior.maxScrollOffset);
-  }
+  void _handleTick() {
+    if (!mounted)
+      return;
 
-  void _handleTabSelected(int tabIndex) {
-    if (_selection != null && tabIndex != _selection.index)
+    if (_controller.indexIsChanging) {
       setState(() {
-        _selection.value = _selection.values[tabIndex];
+        // Rebuild so that the tab label colors reflect the selected tab index.
+        // The first build for a new _controller.index value will also trigger
+        // a scroll to center the selected tab.
       });
+    } else if (config.isScrollable) {
+      _scrollToControllerValue();
+    }
   }
 
-  Widget _toTab(TabLabel label, int tabIndex, Color color, Color selectedColor) {
-    Color labelColor = color;
-    if (_selection != null) {
-      final bool isSelectedTab = tabIndex == _selection.index;
-      final bool isPreviouslySelectedTab = tabIndex == _selection.previousIndex;
-      labelColor = isSelectedTab ? selectedColor : color;
-      if (_selection.valueIsChanging) {
-        if (isSelectedTab)
-          labelColor = Color.lerp(color, selectedColor, _selection.animation.value);
-        else if (isPreviouslySelectedTab)
-          labelColor = Color.lerp(selectedColor, color, _selection.animation.value);
-      }
-    }
-    return new _Tab(
-      onSelected: () { _handleTabSelected(tabIndex); },
-      label: label,
-      color: labelColor
-    );
+  void _saveTabOffsets(List<double> tabOffsets) {
+    if (_indicatorPainter != null)
+      _indicatorPainter.tabOffsets = tabOffsets;
```

Comment:
I think the following will work:

`_indicatorPainter?.tabOffsets = tabOffsets;`